In [1]:
import numpy as np
import pandas as pd

In [2]:
walmart_data = pd.read_csv(r"E:\Simplilearn\project walmart\Walmart_Store_sales.csv")

In [38]:
walmart_data.head()

Store        Date  Weekly_Sales  Holiday_Flag  Temperature  Fuel_Price  \
0      1  05-02-2010    1643690.90             0        42.31       2.572   
1      1  12-02-2010    1641957.44             1        38.51       2.548   
2      1  19-02-2010    1611968.17             0        39.93       2.514   
3      1  26-02-2010    1409727.59             0        46.63       2.561   
4      1  05-03-2010    1554806.68             0        46.50       2.625   

          CPI  Unemployment       Day  
0  211.096358         8.106    Sunday  
1  211.242170         8.106  Thursday  
2  211.289143         8.106    Friday  
3  211.319643         8.106    Friday  
4  211.350143         8.106    Monday

# Basic Statistics 

Which store has maximum sales

In [4]:
wallmart_groupdata = walmart_data.groupby('Store')['Weekly_Sales'].sum()

In [5]:
print("The Store number {} has the maximum sales {}".format(wallmart_groupdata.idxmax(),wallmart_groupdata.max()))

The Store number 20 has the maximum sales 301397792.46


Which store has maximum standard deviation i.e., the sales vary a lot. Also, find out the coefficient of mean to standard deviation


In [6]:
walmart_std= walmart_data.groupby('Store').agg({'Weekly_Sales':'std'})

In [7]:
print('The store number {} has the maximum Standard Deviation, that is {}'.format(walmart_std['Weekly_Sales'].idxmax(),
                                                                                  walmart_std['Weekly_Sales'].max()))

The store number 14 has the maximum Standard Deviation, that is 317569.9494755081


In [8]:
walmart_std1= walmart_data.groupby('Store').agg({'Weekly_Sales':['mean','std']})

In [9]:
walmart_std1.head(10)

Weekly_Sales               
               mean            std
Store                             
1      1.555264e+06  155980.767761
2      1.925751e+06  237683.694682
3      4.027044e+05   46319.631557
4      2.094713e+06  266201.442297
5      3.180118e+05   37737.965745
6      1.564728e+06  212525.855862
7      5.706173e+05  112585.469220
8      9.087495e+05  106280.829881
9      5.439806e+05   69028.666585
10     1.899425e+06  302262.062504

Which store/s has good quarterly growth rate in Q3’2012

In [10]:
wallmartQ3_2012 = walmart_data[(pd.to_datetime(walmart_data['Date']) >= pd.to_datetime('01-07-2012')) & 
                               (pd.to_datetime(walmart_data['Date']) <= pd.to_datetime('30-09-2012'))]

In [11]:
walmart_dataQ3_grp = wallmartQ3_2012.groupby(['Store'])['Weekly_Sales'].sum()

In [12]:
walmartQ3_max = print('The store number {} has achieved the good quality growth {} in third Quarter of 2012'.format
                      (walmart_dataQ3_grp.idxmax(),walmart_dataQ3_grp.max()))

The store number 4 has achieved the good quality growth 77516251.42 in third Quarter of 2012


In [13]:
walmart_data['Weekly_Sales'].mean()

1046964.8775617732

Some holidays have a negative impact on sales. Find out holidays which have higher sales than the mean sales in 
non-holiday season for all stores together


In [14]:
stores_holiday_sales = walmart_data[walmart_data['Holiday_Flag'] ==1]

In [15]:
stores_nonholiday_sales = walmart_data[walmart_data['Holiday_Flag'] ==0]

In [16]:
#Super Bowl: 12-Feb-10, 11-Feb-11, 10-Feb-12, 8-Feb-13

stores_holiday_sales_superBowl = stores_holiday_sales[(pd.to_datetime(stores_holiday_sales['Date'])== pd.to_datetime('12-02-2010')) | 
                                                      (pd.to_datetime(stores_holiday_sales['Date'])== pd.to_datetime('11-02-2011')) |
                                                      (pd.to_datetime(stores_holiday_sales['Date'])== pd.to_datetime('10-02-2012'))|
                                                      (pd.to_datetime(stores_holiday_sales['Date'])== pd.to_datetime('08-02-2013'))]

In [17]:
#Labour Day: 10-Sep-10, 9-Sep-11, 7-Sep-12, 6-Sep-13
stores_holiday_sales_labourDay = stores_holiday_sales[(pd.to_datetime(stores_holiday_sales['Date']) == pd.to_datetime('10-09-2010'))|
                                                     (pd.to_datetime(stores_holiday_sales['Date']) == pd.to_datetime('09-09-2011'))|
                                                     (pd.to_datetime(stores_holiday_sales['Date']) == pd.to_datetime('07-09-2012'))|
                                                     (pd.to_datetime(stores_holiday_sales['Date']) == pd.to_datetime('06-09-2013'))]

In [18]:
#Thanksgiving: 26-Nov-10, 25-Nov-11, 23-Nov-12, 29-Nov-13
stores_holiday_sales_Thanksgiving = stores_holiday_sales[(pd.to_datetime(stores_holiday_sales['Date']) == pd.to_datetime('26-11-2010'))|
                                                        (pd.to_datetime(stores_holiday_sales['Date']) == pd.to_datetime('25-11-2011'))|
                                                        (pd.to_datetime(stores_holiday_sales['Date']) == pd.to_datetime('23-11-2012'))|
                                                        (pd.to_datetime(stores_holiday_sales['Date']) == pd.to_datetime('29-11-2013'))]

In [19]:
#Christmas: 31-Dec-10, 30-Dec-11, 28-Dec-12, 27-Dec-13

stores_holiday_sales_Christmas = stores_holiday_sales[(pd.to_datetime(stores_holiday_sales['Date'])== pd.to_datetime('31-12-2010'))|
                                                     (pd.to_datetime(stores_holiday_sales['Date'])== pd.to_datetime('30-12-2011'))|
                                                     (pd.to_datetime(stores_holiday_sales['Date'])== pd.to_datetime('28-12-2012'))|
                                                     (pd.to_datetime(stores_holiday_sales['Date'])== pd.to_datetime('27-12-2013'))]

In [20]:
stores_nonholiday_sales_mean = stores_nonholiday_sales.groupby('Date').agg({'Weekly_Sales':'mean'}).reset_index()

In [21]:
stores_holiday_sales_sum = stores_holiday_sales.groupby('Date').agg({'Weekly_Sales':'sum'}).reset_index()

In [22]:
stores_holiday_sales_sum

Date  Weekly_Sales
0  07-09-2012   48330059.31
1  09-09-2011   46763227.53
2  10-02-2012   50009407.92
3  10-09-2010   45634397.84
4  11-02-2011   47336192.79
5  12-02-2010   48336677.63
6  25-11-2011   66593605.26
7  26-11-2010   65821003.24
8  30-12-2011   46042461.04
9  31-12-2010   40432519.00

In [23]:
for row in stores_holiday_sales_sum.itertuples():
    for row1 in stores_nonholiday_sales_mean.itertuples():
        if row.Weekly_Sales > row1.Weekly_Sales:
            print("On this Date {} Holiday Sales is greater than Non Holiday Sales and the Sales :- {}".format
                  (row.Date,row.Weekly_Sales))
            break;

On this Date 07-09-2012 Holiday Sales is greater than Non Holiday Sales and the Sales :- 48330059.31
On this Date 09-09-2011 Holiday Sales is greater than Non Holiday Sales and the Sales :- 46763227.53
On this Date 10-02-2012 Holiday Sales is greater than Non Holiday Sales and the Sales :- 50009407.92
On this Date 10-09-2010 Holiday Sales is greater than Non Holiday Sales and the Sales :- 45634397.839999996
On this Date 11-02-2011 Holiday Sales is greater than Non Holiday Sales and the Sales :- 47336192.79
On this Date 12-02-2010 Holiday Sales is greater than Non Holiday Sales and the Sales :- 48336677.63
On this Date 25-11-2011 Holiday Sales is greater than Non Holiday Sales and the Sales :- 66593605.26
On this Date 26-11-2010 Holiday Sales is greater than Non Holiday Sales and the Sales :- 65821003.24
On this Date 30-12-2011 Holiday Sales is greater than Non Holiday Sales and the Sales :- 46042461.04
On this Date 31-12-2010 Holiday Sales is greater than Non Holiday Sales and the Sale

In [24]:
print("Total sales on holidays :-")

print("Super Bowl Day Sale :",stores_holiday_sales_superBowl['Weekly_Sales'].sum())
print("Labour Day Sale :",stores_holiday_sales_labourDay['Weekly_Sales'].sum())
print("Thanksgiving Day Sale :",stores_holiday_sales_Thanksgiving['Weekly_Sales'].sum())
print("Christmas Day Sale :",stores_holiday_sales_Christmas['Weekly_Sales'].sum())

Total sales on holidays :-
Super Bowl Day Sale : 145682278.34
Labour Day Sale : 140727684.68
Thanksgiving Day Sale : 132414608.5
Christmas Day Sale : 86474980.03999999


For Store 1 – Build  prediction models to forecast demand

Linear Regression – Utilize variables like date and restructure dates as 1 for 5 Feb 2010
(starting from the earliest date in order). Hypothesize if CPI, unemployment, and fuel price have any impact on sales.

Change dates into days by creating new variable.

In [25]:
x_features_object = walmart_data[walmart_data['Store'] ==1][['Store','Date']]
date_obj = walmart_data[walmart_data['Store'] ==1][['Date']]
date_obj.index += 1 
x_features_object.Date = date_obj.index
x_features_object

Store  Date
0        1     1
1        1     2
2        1     3
3        1     4
4        1     5
..     ...   ...
138      1   139
139      1   140
140      1   141
141      1   142
142      1   143

[143 rows x 2 columns]

In [26]:
y_target = walmart_data[walmart_data['Store'] ==1]['Weekly_Sales']
y_target

0      1643690.90
1      1641957.44
2      1611968.17
3      1409727.59
4      1554806.68
          ...    
138    1437059.26
139    1670785.97
140    1573072.81
141    1508068.77
142    1493659.74
Name: Weekly_Sales, Length: 143, dtype: float64

# Building prediction models to forecast demand

In [27]:
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test = train_test_split(x_features_object,y_target,random_state=1)

In [28]:
from sklearn.linear_model import LinearRegression
linreg = LinearRegression()
linreg.fit(x_train,y_train)
feature_ds = walmart_data[walmart_data['Store'] ==1][['Store','CPI','Unemployment','Fuel_Price']]
feature_ds.head()

Store         CPI  Unemployment  Fuel_Price
0      1  211.096358         8.106       2.572
1      1  211.242170         8.106       2.548
2      1  211.289143         8.106       2.514
3      1  211.319643         8.106       2.561
4      1  211.350143         8.106       2.625

In [29]:
resp_set_cpi = walmart_data[walmart_data['Store'] ==1]['CPI'].astype('int64')


In [30]:
resp_set_unemployment = walmart_data[walmart_data['Store'] ==1]['Unemployment'].astype('int64')

In [31]:
from sklearn.model_selection import train_test_split
x_train_cpi,x_test_cpi,y_train_cpi,y_test_cpi = train_test_split(feature_ds,resp_set_cpi,random_state=1)
x_train_unemp, x_test_unemp, y_train_unemp, y_test_unemp = train_test_split(feature_ds,resp_set_unemployment,random_state=1)

In [32]:
from sklearn.linear_model import LogisticRegression
logreg = LogisticRegression(max_iter=10000)
logreg.fit(x_train_cpi,y_train_cpi)
y_pred = logreg.predict(x_test_cpi)
logreg.fit(x_train_unemp,y_train_unemp)


LogisticRegression(max_iter=10000)

In [33]:
y_pred_unemp = logreg.predict(x_test_unemp)

In [34]:
y_pred_unemp

array([7, 7, 7, 7, 6, 7, 7, 7, 7, 7, 6, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7,
       7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7], dtype=int64)

In [35]:
from sklearn import metrics
print(metrics.accuracy_score(y_test_cpi,y_pred))
print(metrics.accuracy_score(y_test_unemp,y_pred_unemp))

0.7222222222222222
0.9444444444444444


In [36]:
print('The Actual CPI :', y_test_cpi.values[0:30])
print('Predicted CPI :', y_pred[0:30])
print('actual Unemployment :', y_test_unemp.values[0:30])
print('Predicted Unemployment :', y_pred_unemp[0:30])

The Actual CPI : [215 221 211 211 221 211 210 211 215 217 221 212 216 218 211 210 211 217
 215 211 212 217 221 219 214 211 211 219 215 219]
Predicted CPI : [215 221 211 211 221 211 211 211 215 215 221 211 215 218 211 211 211 217
 215 211 211 217 221 220 215 211 211 221 215 220]
actual Unemployment : [7 7 7 8 7 7 7 7 7 7 6 7 7 7 7 7 7 7 7 7 7 7 7 7 7 7 7 7 7 7]
Predicted Unemployment : [7 7 7 7 6 7 7 7 7 7 6 7 7 7 7 7 7 7 7 7 7 7 7 7 7 7 7 7 7 7]


In [37]:
walmart_data['Day'] = pd.to_datetime(walmart_data['Date']).dt.day_name()
walmart_data.head()

Store        Date  Weekly_Sales  Holiday_Flag  Temperature  Fuel_Price  \
0      1  05-02-2010    1643690.90             0        42.31       2.572   
1      1  12-02-2010    1641957.44             1        38.51       2.548   
2      1  19-02-2010    1611968.17             0        39.93       2.514   
3      1  26-02-2010    1409727.59             0        46.63       2.561   
4      1  05-03-2010    1554806.68             0        46.50       2.625   

          CPI  Unemployment       Day  
0  211.096358         8.106    Sunday  
1  211.242170         8.106  Thursday  
2  211.289143         8.106    Friday  
3  211.319643         8.106    Friday  
4  211.350143         8.106    Monday

## Here the Logistics Regression model is giving the best Accuracy of 94.4%